pgx/src/heap_tuple.rs

```diff
@@ -0,0 +1,340 @@
+        }
+
+        unsafe {
+            let formed_tuple =
+                pg_sys::heap_form_tuple(tupdesc.as_ptr(), datums.as_mut_ptr(), nulls.as_mut_ptr());
+
+            Ok(Self {
+                tuple: PgBox::<pg_sys::HeapTupleData, AllocatedByRust>::from_rust(formed_tuple),
+                tupdesc,
+            })
+        }
+    }
+
+    /// Creates a new [PgHeapTuple] from an opaque Datum that should be a "composite" type.
+    ///
+    /// The Datum should be a pointer to a [pg_sys::HeapTupleHeader].  Typically, this will be used
+    /// in situations when working with SQL `ROW(...)` constructors, or a composite SQL type such as
+    ///
+    /// ```sql
+    /// CREATE TYPE my_composite AS (name text, age i32);
+    /// ```
+    ///
+    /// ## Safety
+    ///
+    /// This function is unsafe as we cannot guarantee that the provided Datum is a valid [pg_sys::HeapTupleHeader]
+    /// pointer.
+    pub unsafe fn from_composite_datum(composite: pg_sys::Datum) -> Self {
+        let htup_header =
+            pg_sys::pg_detoast_datum(composite as *mut pg_sys::varlena) as pg_sys::HeapTupleHeader;
+        let tup_type = crate::heap_tuple_header_get_type_id(htup_header);
+        let tup_typmod = crate::heap_tuple_header_get_typmod(htup_header);
+        let tupdesc = pg_sys::lookup_rowtype_tupdesc(tup_type, tup_typmod);
+
+        let mut data = PgBox::<pg_sys::HeapTupleData>::alloc0();
+
+        data.t_len = crate::heap_tuple_header_get_datum_length(htup_header) as u32;
+        data.t_data = htup_header;
+
+        Self {
+            tuple: data,
+            tupdesc: PgTupleDesc::from_pg(tupdesc),
+        }
+    }
+
+    /// Given the name for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute names are case sensitive.
+    ///
+    /// ## Errors
+    ///
+    /// - return [TryFromDatumError::NoSuchAttributeName] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_name<T: IntoDatum>(&mut self, attname: &str, value: T) -> FromDatumResult<()> {
+        match self.get_attribute_by_name(attname) {
+            None => Err(TryFromDatumError::NoSuchAttributeName(attname.to_string())),
+            Some((attnum, _)) => self.set_by_index(attnum, value),
+        }
+    }
+
+    /// Given the index for an attribute in this [PgHeapTuple], change its value.
+    ///
+    /// Attribute numbers start at 1, not 0.
+    ///
+    /// ## Errors
+    /// - return [TryFromDatumError::NoSuchAttributeNumber] if the attribute does not exist
+    /// - return [TryFromDatumError::IncompatibleTypes] if the Rust type of the `value` is not
+    /// compatible with the attribute's Postgres type
+    pub fn set_by_index<T: IntoDatum>(
+        &mut self,
+        attno: NonZeroUsize,
+        value: T,
+    ) -> FromDatumResult<()> {
```

Comment:
It strikes me that this we could have some of this function extracted into a `replace_at_index(..) -> Result<Self, _>` which returns the new heaptuple, and call it here as well as exposing that.

Reply:
I don't follow?

Reply:
Given that this function always must create a copy, it might desirable to support a pattern like:

```rust
let cloned = my_tuple.replace_at_index(1, "thing");
```

Instead of

```rust
let cloned = my_tuple.clone();
cloned.set_at_index(1, "thing");
```

This would allow the user to avoid an extra clone happening. (and this function could call that one, so it's not added complexity on our side)

Reply:
PgHeapTuple doesn’t implement Clone so I still don’t follow…